dali/util/fits.cc

```diff
@@ -0,0 +1,135 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/util/fits.h"
+#include <fitsio.h>
+#include <string>
+#include <vector>
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+namespace fits {
+
+
+std::string GetFitsErrorMessage(int status) {
+  std::string status_str;
+  status_str.resize(FLEN_STATUS);
+
+  fits_get_errstatus(status, &status_str[0]); /* get the error description */
+
+  return status_str;
+}
+
+void HandleFitsError(int status) {
+  if (status) {
+    DALI_FAIL(GetFitsErrorMessage(status));
+  }
+}
+
+int ImgTypeToDatatypeCode(int img_type) {
+  switch (img_type) {
+    case SBYTE_IMG:
+      return TSBYTE;
+    case BYTE_IMG:
+      return TBYTE;
+    case SHORT_IMG:
+      return TSHORT;
+    case USHORT_IMG:
+      return TUSHORT;
+    case LONG_IMG:
+      return TINT;
+    case ULONG_IMG:
+      return TUINT;
+    case LONGLONG_IMG:
+      return TLONGLONG;
+    case ULONGLONG_IMG:
+      return TULONGLONG;
+    case FLOAT_IMG:
+      return TFLOAT;
+    case DOUBLE_IMG:
+      return TDOUBLE;
+    default:
+      DALI_FAIL("Unknown BITPIX value! Refer to the CFITSIO documentation.");
+  }
+}
+
+const TypeInfo &TypeFromFitsDatatypeCode(int datatype) {
+  switch (datatype) {
+    case TSBYTE:
+      return TypeTable::GetTypeInfo<int8_t>();
+    case TBYTE:
+      return TypeTable::GetTypeInfo<uint8_t>();
+    case TSHORT:
+      return TypeTable::GetTypeInfo<int16_t>();
+    case TUSHORT:
+      return TypeTable::GetTypeInfo<uint16_t>();
+    case TINT:
+      return TypeTable::GetTypeInfo<int32_t>();
+    case TUINT:
+      return TypeTable::GetTypeInfo<uint32_t>();
+    case TLONGLONG:
+      return TypeTable::GetTypeInfo<int64_t>();
+    case TULONGLONG:
+      return TypeTable::GetTypeInfo<uint64_t>();
+    case TFLOAT:
+      return TypeTable::GetTypeInfo<float>();
+    case TDOUBLE:
+      return TypeTable::GetTypeInfo<double>();
+    default:
+      DALI_FAIL("Unknown datatype code value! Refer to the CFITSIO documentation.");
+  }
+}
+
+void ParseHeader(HeaderData &parsed_header, fitsfile *src) {
+  int32_t hdu_type, img_type, n_dims, status = 0;
+
+  fits_get_hdu_type(src, &hdu_type, &status);
+  bool is_image = (hdu_type == IMAGE_HDU);
+  DALI_ENFORCE(is_image, "Only IMAGE_HDUs are supported!");
+
+  fits_get_img_equivtype(src, &img_type, &status); /* get IMG_TYPE code value */
+  fits_get_img_dim(src, &n_dims, &status);         /* get NAXIS value */
+
+  DALI_ENFORCE(n_dims > 0, "NAXIS (image dimensions) value for each HDU has to be greater than 0!");
```

Comment:
I think we still need an extra `HandleFitsError` check here (or `n_dims=0`) in the definition. Otherwise, if `fits_get_img_dim` fails, we DALI_ENFORCE with something unitilizied. It could be positive but still garabge.

Reply:
Yup, although I went ahead and simply wrapped everything in FITS_CALL.

Reply:
Cool!